Introducing extension mechanism to load custom DB in Routerlicious

nedalhy opened the pull request:
This PR introduces extension loading mechanism for using custom DB in Routerlicious, while maintaining backward compatibility.

It relies on the config file to determine which db should it connect to, and then dynamically importing the corresponding DB extension and instantiates the factory object.

```{
   "loadExtensions": true 
    "db" : { 
          "name" : "SomeDb" ,
    },
    "extensions": {
       "db": {
           "modulePathOrName": "PATH_TO_EXTENSION_BUNDLE",
           "dbConfig": {.....}  // Custom configuration,
           "name": "SomeDb" // The DB name,
           "dbFactoryConstructorName": "SomeDbFactory" // The factory class name 
        }
    }
}
```

karlbom committed:
make database configurable and extendable

nedalhy commented:
@evaliyev @ruiterr PTAL

znewton commented on server/routerlicious/packages/services-core/src/dbFactoryFactory.ts:
```diff
@@ -0,0 +1,59 @@
+            dbServices.map((desc) => [desc.name, desc.factory]),
+        );
+
+        if (config.get('loadExtensions')) {
```
I think this change would overall make more sense if, instead of using a super generic concept like "extensions," we were to just make a single "db" config. Perhaps I am missing something, but I don't see a need for having the additional complexity of loading and configuring several database modules.

nedalhy replied:
Yes I agree that it adds additional complexity. 

The idea is to provide some flexibility for the users, to load custom DB drivers which are not part of the Fluid repo (e.g. DynamoDB driver). Additionally, to introduce some infrastructure in place to allow adding other extensions in the future, which are not necessary for db.

znewton commented on server/routerlicious/packages/services/src/dbFactory.ts:
```diff
@@ -0,0 +1,16 @@
+import { DbFactoryFactory } from "@fluidframework/server-services-core";
+import { Provider } from "nconf";
+import { MongoDbFactory } from "./mongodb";
+export class RouterlicousDbFactoryFactory extends DbFactoryFactory {
```
Proposal: instead of adding a FactoryFactory, you could just have a function that returns a Factory.

```ts
interface IDBFactoryConfig {
  name: string;
  modulePathOrName: string;
  dbFactoryConstructorName: string;
  dbConfig: any;
}
export function getDbFactory(config: Provider): IDBFactory {
  const dbFactoryConfig = config.get("db");
  if (!dbFactoryConfig || dbFactoryConfig.name === "mongo") {
    // Taken from current ResourceFactories
    const mongoUrl = config.get("mongo:operationsDbEndpoint") as string;
    const bufferMaxEntries = config.get("mongo:bufferMaxEntries") as number | undefined;
    const mongoFactory = new services.MongoDbFactory(mongoUrl, bufferMaxEntries);
    return mongoFactory;
  }
  const externalDbFactoryModule = require(dbFactoryConfig.modulePathOrName);
  const externalDbFactory = new externalDbFactoryModule[dbFactoryConfig.dbFactoryConstructorName](
    dbFactoryConfig.dbConfig,
  ) as IDBFactory;
  return externalDbFactory;
}
```

I believe you wouldn't need any DBFactoryFactories if you used this or something very similar. And it should maintain back-compat perfectly fine.

nedalhy committed:
Addressing comments

nedalhy commented on server/routerlicious/packages/services/src/dbFactory.ts:
```diff
@@ -0,0 +1,48 @@
+        const extension = EXTENSIONS.find((ext) => ext.name === dbFactoryConfig.name);
+
+        if (extension === undefined) {
+            throw Error(`The database ${dbFactoryConfig.name} is not found.`);
```
It should have the same effect, but I can change it if you prefer. 
Both versions are used across the repo.

nedalhy commented on server/routerlicious/packages/services/src/mongodb.ts:
```diff
@@ -155,11 +156,29 @@ export class MongoDb implements core.IDb {
 
-    public async connect(): Promise<core.IDb> {
+    public async connect(global = false): Promise<core.IDb> {
+        assert(global && !this.globalDbEndpoint, `No global endpoint provided when trying to connect to global db.`);
```
Right, nice catch.

nedalhy committed:
Address comments 1

znewton commented on server/routerlicious/packages/services-core/src/mongo.ts:
```diff
@@ -4,59 +4,49 @@
 /**
- * Helper class to manage access to a MongoDb database
+ * Helper class to manage access to database
+ * @TODO: Rename the file name as it behaves now as a generic DB Manager
```
curious: what is the blocker for doing this rename in this PR? I think renaming to `DatabaseManager` or `DbManager` is a great idea

nedalhy replied:
Renaming the class and the file would make it harder to review the the core changes of this PR due to the number of changes. Additionally, it would increase the difficulty of rebase for other pending PRs after its merged.

My plan was to address renaming related changes in a dedicated PR. But if you still prefer to address all of them on this PR, I can include these changes too.

pradeepvairamani commented on server/routerlicious/packages/services/src/dbFactory.ts:
```diff
@@ -0,0 +1,55 @@
+
+    // We enable extensions loading only when the flag is set to true.
+    // The idea behind the flag to have an additional security measure to avoid loading unwanted extensions.
+    // @TODO: in the initial proposal it was part of the config file, but perhaps it should be an
```
I am curious about this Todo item. Why might we want this to be an env variable instead of a config as it is now?

pradeepvairamani commented:
Another general comment: How was this PR tested? By running clicker against it?

hedasilv commented on server/routerlicious/packages/services-core/src/mongo.ts:
```diff
@@ -89,18 +79,18 @@ export class MongoManager {
-        debug("Successfully connected to MongoDB");
-        Lumberjack.info("Successfully connected to MongoDB");
+        debug(`Successfully connected`);
+        Lumberjack.info(`Successfully connected to Db`);
```
Nit: why change the quote style here (using \` now) if the original style is still used above (e.g. line 38, `"`)? Should we just keep everything as `"`?

nedalhy replied:
I used to have a parameter in the template to indicate which DB is currently used (was useful for debugging), I removed it and forgot to change it back to use `"`.

hedasilv commented on server/routerlicious/packages/services/src/dbFactory.ts:
```diff
@@ -0,0 +1,55 @@
+}
+
+const creator = (dbFactoryConfig: IDBFactoryConfig) => async (): Promise<IDbFactory> => {
+    debug(`Using ${dbFactoryConfig.name} Database`);
```
@znewton @pradeepvairamani do you know if we get to call `dbFactory` from client code? Since this `dbFactory` is placed under the services package, I think we can assume it's Node.js exclusively, right?

If that is the case, can we use `Lumberjack` instead of `debug`?
More info about Lumberjack available here: https://github.com/microsoft/FluidFramework/blob/main/server/routerlicious/packages/services-telemetry/README.md

znewton replied:
Correct yes. I think `debug` is a relic here and probably outside the scope of this change. But bonus kudos to @nedalhy if you wanna change it!

nedalhy replied:
Yes sure, I can change it.

hedasilv commented on server/routerlicious/packages/services/src/dbFactory.ts:
```diff
@@ -0,0 +1,55 @@
+    // @TODO: in the initial proposal it was part of the config file, but perhaps it should be an
+    // environment variable instead ?
+    if (config.get("loadExtensions") === true) {
+        const EXTENSIONS = config.get("extensions:db") as IDBFactoryConfig[] || [];
```
nit: any reason to use all capital letters for `EXTENSIONS`? why not just `extensions`?

hedasilv commented on server/tinylicious/src/services/levelDb.ts:
```diff
@@ -63,3 +63,16 @@ export class LevelDb extends EventEmitter implements IDb {
+    }
+
+    public async connect(): Promise<IDb> {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Curious, why disable this eslint rule? Instead, doesn't

`private readonly db: LevelDb;`

solve the issue?

hedasilv commented on server/tinylicious/src/services/inMemorydb.ts:
```diff
@@ -23,3 +23,16 @@ export class InMemoryDb extends EventEmitter implements IDb {
+    }
+
+    public async connect(): Promise<IDb> {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Same here: why disable this eslint rule? Instead, doesn't

private readonly db: LevelDb;

solve the issue?

tianzhu007 commented:
I try to make the `globalDbEnabled` as **true** in the config.json in the r11s sever, and run the routerlicious on that. However, the application doesn't work fine for it. Did you have the trial for that? Thanks. 
@nedalhy [nedalhy](https://github.com/nedalhy)

nedalhy commented:
> 



> tried to make the `globalDbEnabled` as **true** in the config.json in the r11s sever for this branch and run the routerlicious on that. However, the application doesn't work fine for it. Did you have the trial for that? Thanks.
> @nedalhy

I just tried the branch after rebase, clicker + another example are working with `globalDbEnabled=true` . 

![image](https://user-images.githubusercontent.com/17700619/164000986-c05dc560-5b95-455f-b1ad-6daeb49de65c.png)

nedalhy committed:
Address comments

nedalhy commented:
@znewton I addressed most of the comments, except file renaming which would be addressed in next PR if that's okay with  you. 

Can you help me merging it ?

znewton commented:
Thanks, @nedalhy!
